### ISSUE TITLE
Inconsistency in initializing empty arrays and dictionaries

### ISSUE THREAD

**Issue by samkim102** [raywenderlich/swift-style-guide#269]
I would have to say, I agree with most of what is stated in the guidelines, but this one seems inconsistent:

https://github.com/raywenderlich/swift-style-guide#type-annotation-for-empty-arrays-and-dictionaries

The guide indicates:
```swift
// preferred
var names: [String] = []
var lookup: [String: Int] = [:]
```
is preferred over
```swift
// not preferred
var names = [String]()
var lookup = [String: Int]()
```

But the guide usually prefers inferred types in the interest of compact code:
```swift
// preferred
let message = "Click the button"
let currentBounds = computeViewBounds()
```
```swift
// not preferred
let message: String = "Click the button"
let currentBounds: CGRect = computeViewBounds()
```

If we were being consistent, shouldn't we always prefer inferred types over explicit declaration?  I would think this is more consistent
```swift
// should be preferred
var names = [String]();
var lookup = [String: Int]();
```

**Comment by RobertGummesson** [comment]
In my opinion, the guide is correct here. It simply says, when required, specify type. When not, drop it.

Ok, but isn't this really a choice of style? No, not really. Let's have a look at some examples:
```
// preferred
var colors: [UIColor]?
var colors: [UIColor] = []
var colors: [UIColor] = [.red, .green, .blue]
```
Now, let's convert those 3 examples into your suggested preference.
```
var colors: [UIColor]?
var colors = [UIColor]()
var colors = [UIColor.red, UIColor.green, UIColor.blue]
```
See what happened there? For the last row, we now need to repeat UIColor for every element in the array. You may also argue that the declaration for the last two are less consistent with the declaration of the optional. 

What about a method then? Well, you could of course choose to always specify type for the sake of being consistent. However, I would imagine that most developers would want to do away with redundant code.
```
// do we need to specify type?
var colors: [UIColor] = rainbowColors()
// no we don't
var colors = rainbowColors()
```
Does that make sense?

**Comment by samkim102** [comment]
I would have gone with
```swift
var colors: [UIColor]?
var colors = [UIColor]()
var colors: [UIColor] = [.red, .green, .blue]
```
Instead of what you did there.  Are we not preferring to use inferred type except when explicit declaration is required?  In that sense the third line seems consistent.

**Comment by RobertGummesson** [comment]
You are right in the sense that I over simplified the rule. Having that said, there's nothing wrong with either approaches, it just comes down to preference. In my case, I find the below more consistent, more readable and more practical when going back and forth empty / non-empty states. I think we should let others chime in.

```
// preferred
var colors: [UIColor]?
var colors: [UIColor] = []
var colors: [UIColor] = [.red]
```

```
// not preferred (in my opinion)
var colors: [UIColor]?
var colors = [UIColor]()
var colors: [UIColor] = [.red]
```

**Comment by JessyCatterwaul** [comment]
@RobertGummesson, you're only taking arrays into consideration. If we only had arrays to populate with array literals, both solutions would be fine. I even used to use what you were suggesting. But then we got sets, and I had to standardize on a convention that included them too. The one you're suggesting is uglier than explicit typing.

```swift
let ints: Set = [1, 1, 2, 2]
let ints2 = Set([1, 1, 2, 2])

var ints: Set<Int> = []
var ints = Set<Int>([])
```
So, after settling on the top choice being better for sets, we backform our array convention to match it.
```swift
var ints = [1, 1, 2, 2]
var ints: [Int] = []
```
